### PR TITLE
Add Notification Inline

### PR DESCRIPTION
**Pull request by amelako**
Addresses #792.

## Purpose

Introduces the new NotificationInline component.
The Notification Inline component non-disruptively provides quick and contextual inline notification and it is integrated within the content section.
Screenshot 2022-01-14 at 15 08 07

It comes in 4 variants, "info", "confirm", "notify" and "alert".
To communicate a message always provide a body copy, and if needed an optional headline can be included above the body copy and an action button.
Inline notifications are part of the content and should be placed near their related part of the content.
Inline notifications can be dismissed by clicking on a close button in the top-right corner.

## Approach and changes

Refer to the Storybook for further implementation and usage guidelines, and to the threads below for more details and discussion on the implementation.

## Definition of done

* [ ] Development completed
* [ ] Reviewers assigned
* [ ] Unit and integration tests
* [ ] Meets minimum browser support
* [ ] Meets accessibility requirements

### CODE REVIEW

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+### Notification Inline variants
+
+<Story id="notification-notificationinline--variants" />
+
+- Use the 'Info' variant for informational, neutral messages.
+- Use the 'Confirm' variant the successful messages.
+- Use the 'Notify' variant for warning messages, and other information a user should be aware of.
+- Use the 'Alert' variant for error messages.
+
+### Inline with headline and action button
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:33]: This story doesn't have an action button, I think it's a good idea to show it.

Maybe we can also have a short text here explaining that this component can be used instead of a toast if. it needs to be interactive

File: packages/circuit-ui/components/NotificationInline/NotificationInline.tsx [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+      onClose: (event: ClickEvent) => void;
+      /**
+       * Text label for the close button for screen readers.
+       * Important for accessibility.
+       */
+      closeButtonLabel: string;
+    }
+  | { onClose?: never; closeButtonLabel?: never };
+
+export type BaseProps = HTMLAttributes<HTMLDivElement> & {
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:51]: Can we share some of these props between the toast and this component? There should be few differences

File: packages/circuit-ui/components/NotificationInline/NotificationInline.tsx [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+  display: flex;
+  flex-direction: column;
+  align-items: flex-start;
+  padding-right: ${theme.spacings.peta};
+  padding-left: ${theme.spacings.mega};
+`;
+
+const Content = styled('div')(contentStyles);
+
+const actionButtonStyles = ({ theme }: StyleProps & ButtonProps) =>
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:139]: Could we use a secondary button and remove its border instead of the custom button here? Then we wouldn't need to rewrite all of the state-specific styles

File: packages/circuit-ui/components/NotificationInline/NotificationInline.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      />
+    ))}
+  </StackInlineMessages>
+);
+
+Variants.args = {
+  body: 'This is a toast message',
+} as NotificationInlineProps;
+
+export const Base = (args: NotificationInlineProps) => (
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.stories.tsx:62]: I think you need to move this story up, because right now the docs is showing the props for the `Variants` story instead of `Base`

File: packages/circuit-ui/components/NotificationInline/NotificationInline.tsx [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+const iconMap = {
+  info: Info,
+  confirm: Confirm,
+  alert: Alert,
+  notify: Notify,
+};
+
+const inlineWrapperStyles = () => css`
+  display: inline-flex;
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:100]: In the end I think that this doesn't work: I played around with the stories to create a more "real-life" scenario, where the notification is on top of a block of text (could also be a card or something):

Screenshot 2022-01-17 at 11 35 28

With `inline` (or `inline-flex`), it keeps the width of the content and there would be no way (as far as I can see) to make it stretch without overriding the component's CSS.

Instead we probably want something like this in most cases:

Screenshot 2022-01-17 at 11 37 27

So I suggest we remove this line. If developers still need to make the notification fit the content, they can also do it by adding a wrapper, which I think is cleaner than overriding styles.

Comment (robinmetral) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:100]: _Side note:_

Also, my related suggestion in #1312 is probably not the best choice: just like the `NotificationInline`, it should stretch in some cases according to the specs

Screenshot 2022-01-17 at 11 42 04

I'll check with design what the expected visual is when it stretches _and_ has an asset (because that's the scenario when it looks unbalanced right now). I'll post an update in the issue at #1312

**Review by robinmetral — COMMENTED** [review]
I like the direction this is going!

Added a few general comments on the component and the docs for now, I'll also re-review later.

For docs I'm kind of missing something that explains what the difference between the toast and this component is, in terms of behavior, a11y etc. They're pretty similar and I expect that some developers will find it confusing. Could we add this somewhere (in the intro maybe)?

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+<Status.Stable />
+
+<Story id="notification-notificationinline--base" />
+<Props />
+
+The Notification Inline component non-disruptively provides quick and contextual inline notification and it is integrated within the content section.
+Inline notifications are frequently used to provide additional guidance to the user, like positive feedback, confirmation about an action or errors in forms, cards or dialog boxes.
+
+## Usage guidelines
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:14]: I think I mentioned this already, but I've been moving away from these vague "usage guidelines" in the docs I've written lately. I think the information is valuable but could be better organized, so for example:

- details about props ("provide body copy", "optional headline or action button"): these can be removed altogether, and we can just make sure to have enough stories that show the different combinations
- accessibility guidelines ("be concise, specific in copy", "place near the content"): this could go in an _Accessibility_ section with reference to WCAG SCs
- variant-specific guidelines ("give clear next steps" when requiring user action): we could add these under the variant itself

So basically, I think that everything we've been putting under "usage guidelines" could be either removed or put elsewhere, in a clearer way—right no it's not obvious to me what these are actually meant for, since they're a pretty subjective mix of guidelines

File: packages/circuit-ui/components/NotificationInline/NotificationInline.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
+/* eslint-disable array-callback-return */
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.stories.tsx:1]: Let's not disable things for the entire file, but for specific lines (or sections) instead. It makes it more explicit

**Review by robinmetral — COMMENTED** [review]
Oops I submitted the previous review by mistake and didn't have time to finish writing my thoughts down

File: packages/circuit-ui/components/NotificationInline/NotificationInline.tsx (hunk 5 above)
Comment (amelako) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:100]: So you suggest to leave it as is? And for banner too, with being stretched in some cases. And leave to the teams to adjust if necessary?

File: packages/circuit-ui/components/NotificationInline/NotificationInline.tsx (hunk 3 above)
Comment (amelako) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:139]: Hmm, in either case I need to add the changes I have now, cause they are all the ones we need for the cta button, changing to bold, changing color and adding underline.

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx (hunk 6 above)
Comment (amelako) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:14]: Thanks a lot. I will adjust it and move the guidelines to the respective sections then.

File: packages/circuit-ui/components/NotificationInline/NotificationInline.tsx (hunk 3 above)
Comment (robinmetral) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:139]: Yes, but we wouldn't have to handle the `hover`, `active` and `focus` states. But I'm fine with keeping it as-is as well, doesn't make a huge difference. If we reuse the cta button elsewhere though I'd extract it into a component (an internal one)

File: packages/circuit-ui/components/NotificationInline/NotificationInline.tsx (hunk 5 above)
Comment (robinmetral) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.tsx:100]: Yes for the inline notification 🙂 as I mentioned, it's going to be easier to constrain the width than to stretch a component that's constrained by default, so I think it's a better solution. What do you think @amelako/@connor-baer?

For the banner I reached out to design like I said and we'll see what they suggest, it's a similar situation with the added challenge of having an optional asset that can make the banner look unbalanced. But let's take that discussion to #1312

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx (hunk 6 above)
Comment (amelako) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx]: Updated the docs, and removed the parts from the Usage guidelines to different sections. Let me know if you think we should do some more changes. Thanks.

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+<Story id="notification-notificationinline--variants" />
+
+- Use the 'Info' variant for informational, neutral messages.
+- Use the 'Confirm' variant the successful messages.
+- Use the 'Notify' variant for warning messages, and other information a user should be aware of.
+- Use the 'Alert' variant for error messages. Be descriptive and give users clear next steps.
+
+### Inline with headline and action button
+
+<Story id="notification-notificationinline--dismissable" />
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:27]: Right now this is kind of mixing up the optional headline/action (name of the subsection) with the "dismissable" use case (name of the story).

Would it be clearer if we split this "component variations" section into three parts?

- `##` Component variations
  - `###` Notification variants
    - variants story
    - text about when to use each variant
  - `###` With headline and action
    - "with headline and action" story (👈 similar to the current "Dismissable" but doesn't need to be dismissable to avoid showing too many things at once)
    - text about using the optional props
  - `###` Dismissable
    - dismissable story (👈 similar to the current "Dismissable" but doesn't need to have a headline and action to avoid showing too many things at once)
    - text about adding `onClose` _and_ a `closeButtonLabel` to make the notification optionally dismissable

WDYT?

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+## Accessibility guidelines
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:29]:
```suggestion
## Accessibility
```

Nitpick for consistency with other pages

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+- Be concise, specific and short in the message.
+- Inline notifications are part of the content and should be placed near their related part of the content.
+- Do not cover other content with inline notifications.
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:33]: I think this is quite obvious, but maybe it's just me

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+- In case you need to provide an action button, use NotificationInline component. Toast messages should not contain any interactive elements. Screen readers will announce raw text within the toast as it appears, without structure and without moving focus to the toast element.
+- Inline notifications should not be added to the page dynamically. For this use case, use Toast notifications, since only dynamic alerts include aria live attributes.
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:35]: Hmmmm this is complicated, because for example if we need to show server-side form submission errors, it's actually better to avoid the toast and use the inline (since the toast violates 2.2.1 timing adjustable)

Ideally developers who need to write a dynamic form submission notification use the `NotificationInline` but implement the live region themselves.

In the future we'll want to package this into form patterns, but in the meantime I'm not sure what the best approach is here, for documentation purposes. I'm just worried that right now we seem to be pointing from the `NotificationInline` to the toasts (because no live regions) and from the toasts back here (because no timing adjustable)—which could be quite confusing

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+### Inline with headline and action button
+
+<Story id="notification-notificationinline--dismissable" />
+
+## Accessibility guidelines
+
+- Be concise, specific and short in the message.
+- Inline notifications are part of the content and should be placed near their related part of the content.
+- Do not cover other content with inline notifications.
+- In case you need to provide an action button, use NotificationInline component. Toast messages should not contain any interactive elements. Screen readers will announce raw text within the toast as it appears, without structure and without moving focus to the toast element.
```
Comment (robinmetral) [inline packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:34]: Maybe we should mention this on the `NotificationToast` docs only? Not sure it makes a lot of sense to add it here; the reader is already on the `NotificationInline` docs

**Review by robinmetral — COMMENTED** [review]
Just a few extra thoughts about docs, thanks for addressing my previous review's more general comments 🙌

File: packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx (hunk 8 above)
Comment (amelako) [reply packages/circuit-ui/components/NotificationInline/NotificationInline.docs.mdx:27]: Sounds good. I like the suggestion.

File: .changeset/nine-tigers-approve.md [hunk 13]
```diff
@@ -0,0 +1,5 @@
+---
+'@sumup/circuit-ui': minor
+---
+
+Added a new `NotificationInline` component that provides quick and contextual inline notification.
```
Comment (robinmetral) [inline .changeset/nine-tigers-approve.md:5]:
```suggestion
Added a new `NotificationInline` component that provides quick and contextual inline notifications.
```

**Review by robinmetral — APPROVED** [review]
Nice! Let's release this and I might add a bit more to the docs separately 👀 The new component looks great!
